src/ObjWriting/XModel/Export/XModelBinWriter.cpp

```diff
@@ -0,0 +1,484 @@
+                          std::numeric_limits<int16_t>::max());
+    }
+
+    uint8_t ClampFloatToUByte(const float value) const
+    {
+        return std::clamp(static_cast<uint8_t>(std::numeric_limits<uint8_t>::max() * std::clamp(value, -1.0f, 1.0f)),
+                          std::numeric_limits<uint8_t>::min(),
+                          std::numeric_limits<uint8_t>::max());
+    }
+
+    void WriteHeader(int16_t version)
+    {
+        WriteComment("OpenAssetTools XMODEL_BIN File");
+        WriteComment(std::format("Game Origin: {}", m_game_name));
+        WriteComment(std::format("Zone Origin: {}", m_zone_name));
+        Write(XModelBinHash::MODEL);
+        WriteInt16(static_cast<int16_t>(XModelBinHash::VERSION), version);
+    }
+
+    void WriteBones(const XModelCommon& xmodel)
+    {
+        WriteInt16(static_cast<int16_t>(XModelBinHash::BONE_COUNT), static_cast<int16_t>(xmodel.m_bones.size()));
+
+        auto boneNum = 0;
+        for (const auto& bone : xmodel.m_bones)
+        {
+            Write(XModelBinHash::BONE);
+            Write(boneNum);
+            if (bone.parentIndex)
+                Write(static_cast<int32_t>(*bone.parentIndex));
+            else
+                Write(-1);
+
+            WriteAlignedString(bone.name);
+            boneNum++;
+        }
+
+        boneNum = 0;
+        for (const auto& bone : xmodel.m_bones)
+        {
+            WriteInt16(static_cast<int16_t>(XModelBinHash::BONE_INDEX), boneNum);
+
+            Write(XModelBinHash::OFFSET);
+            Write(bone.globalOffset[0]); // X
+            Write(bone.globalOffset[1]); // Y
+            Write(bone.globalOffset[2]); // Z
+
+            Write(XModelBinHash::BONE_SCALE);
+            Write(bone.scale[0]); // X
+            Write(bone.scale[1]); // Y
+            Write(bone.scale[2]); // Z
+
+            const auto mat = Eigen::Quaternionf(bone.globalRotation.w, bone.globalRotation.x, bone.globalRotation.y, bone.globalRotation.z).matrix();
+
+            Write(static_cast<int16_t>(XModelBinHash::BONE_MATRIX_X));
+            Write(ClampFloatToShort(mat(0, 0)));
+            Write(ClampFloatToShort(mat(0, 1)));
+            Write(ClampFloatToShort(mat(0, 2)));
+
+            Write(static_cast<int16_t>(XModelBinHash::BONE_MATRIX_Y));
+            Write(ClampFloatToShort(mat(1, 0)));
+            Write(ClampFloatToShort(mat(1, 1)));
+            Write(ClampFloatToShort(mat(1, 2)));
+
+            Write(static_cast<int16_t>(XModelBinHash::BONE_MATRIX_Z));
+            Write(ClampFloatToShort(mat(2, 0)));
+            Write(ClampFloatToShort(mat(2, 1)));
+            Write(ClampFloatToShort(mat(2, 2)));
+
+            boneNum++;
+        }
+    }
+
+    XModelBinWriterBase(std::ostream& stream, std::string gameName, std::string zoneName)
+        : m_stream(stream),
+          m_game_name(std::move(gameName)),
+          m_zone_name(std::move(zoneName))
+    {
+    }
+
+    std::ostringstream m_writer;
+    std::ostream& m_stream;
```

Comment:
The `m_stream` variable is not used in the parent class, consider moving it to the child class.